### PR TITLE
Stop using controller storage

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
- Controller storage is unreliable: in a model migration we once saw `ops.storage.NoSnapshotError: no snapshot data found for LokiOperatorCharm/LokiPushApiProvider[logging]/on/loki_push_api_alert_rules_changed[250784] object`, which broke the upgrade.
- Controller storage is [deprecated in ops](https://github.com/canonical/operator/pull/882).

## Solution
Stop using controller storage.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -128,9 +128,9 @@ def __init__(self, *args):
                 config=to_tuple(ActiveStatus()),
                 rules=to_tuple(ActiveStatus()),
                 retention=to_tuple(ActiveStatus()),
-            )
+            ),
+            fresh_install=True,
         )
-        self._stored.set_default(fresh_install=True)  # Relies on controller-backed storage
```
Comment (sed-i) [inline src/charm.py:134]: @IbraAoad: we initially said this relies on controller storage, but now that I see we set `fresh_install` it to False and the config backup is saved to persistent storage, it seems that controller storage isn't necessary. Unless I'm missing something?

File: src/charm.py [hunk 2]
```diff
@@ -793,6 +793,4 @@ def server_ca_cert_path(self) -> Optional[str]:
 
 
 if __name__ == "__main__":
-    # We need use_juju_for_storage=True because ingress_per_unit
-    # requires it to keep track of the ingress address.
-    main(LokiOperatorCharm, use_juju_for_storage=True)
```
Comment (sed-i) [inline src/charm.py:798]: @PietroPasotti @dstathis Our other IPU charm, prometheus, doesn't use controller storage.
And in ingress_per_unit v1, the only use of stored state is for current_urls, which seems like we don't need to rely on controller storage. Wdyt?

Comment (PietroPasotti) [reply src/charm.py:798]: I'd tend to agree. worst that can happen is you get two ready events instead of one, and given the hot reloading this won't even trigger a restart.

File: src/charm.py [hunk 3]
```diff
@@ -128,9 +128,9 @@ def __init__(self, *args):
                 config=to_tuple(ActiveStatus()),
                 rules=to_tuple(ActiveStatus()),
                 retention=to_tuple(ActiveStatus()),
-            )
+            ),
+            fresh_install=True,
```
Comment (PietroPasotti) [inline src/charm.py:132]: it's not clear to me what this does, and why this change was necessary. I realize this is hard to test as we'd need an integration test to redeploy the charm, or simulate pod churn? 
But at least let's add a detailed comment to explain what's going on.

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
we need a little bit more comments/ documentation to explain to our future selves what this is all about

File: src/charm.py (hunk 3 above)
Comment (PietroPasotti) [reply src/charm.py:132]: if you do think this is testable though, please let's write a test

**Review by lucabello — APPROVED** [review]

File: src/charm.py (hunk 3 above)
Comment (sed-i) [reply src/charm.py:132]: It was introduced in #393 and there's an itest.

File: src/charm.py (hunk 1 above)
Comment (ca-scribner) [reply src/charm.py:134]: yeah I think I agree with this.  Persisting data between pods due to upgrade, churn, etc, shouldn't be needed.  This flag I think is only meaningful within a given pod's lifecycle.  It says whether this pod started its life in an install or upgrade, and I think that's all that's necessary

File: src/charm.py (hunk 3 above)
Comment (ca-scribner) [reply src/charm.py:132]: Yeah this code is from a previous commit, but like @PietroPasotti I found it tough to understand the importance at first.  Maybe a comment or link to an issue explaining it could help.  Although it does get better explanation in the relevant part of `configure` further down

**Review by ca-scribner — APPROVED** [review]

File: src/charm.py (hunk 3 above)
Comment (sed-i) [reply src/charm.py:136]: Yep, @IbraAoad documented this in two places next to where it's used:
- https://github.com/canonical/loki-k8s-operator/blob/fd3e58407ef55c5150ac5b54bf42dd4fb8a7dadc/src/charm.py#L492
- https://github.com/canonical/loki-k8s-operator/blob/fd3e58407ef55c5150ac5b54bf42dd4fb8a7dadc/src/charm.py#L252

**Comment by lucabello** [comment]
Failing integration tests after the latest commit :(

**Comment by sed-i** [comment]
> Failing integration tests after the latest commit :(

Got rate limited

```
unknown container reason "ImagePullBackOff": Back-off pulling image "docker.io/jujusolutions/charm-base:ubuntu-20.04"
```
